database/src/main/java/database/dao/mysql/MySQLConnectionPool.java

```diff
@@ -92,34 +100,46 @@ public Connection getConnection() {
     /**
      * Releases a connection back into the pool. If the pool is full, the connection is closed.
      *
+     * <p>This method is thread-safe, ensuring that multiple threads can safely return
+     * connections to the pool without risking race conditions.
+     *
      * @param connection the {@link Connection} to be released
      */
     public void releaseConnection(Connection connection) {
-        if (connection != null && connectionPool.size() < maxPoolSize) {
-            connectionPool.add(connection);
-        } else if (connection != null) {
-            closeConnection(connection);
+        if (connection != null) {
+            synchronized (this) {
+                if (connectionPool.size() < maxPoolSize) {
+                    connectionPool.add(connection);
+                } else {
+                    closeConnection(connection);
+                }
+            }
         }
     }
 
     /**
      * Closes all connections in the pool and resets the current connection count.
      * This method should be called when the pool is no longer needed, for example
      * during application shutdown, to release resources.
+     *
+     * <p>This method is thread-safe, allowing safe invocation in concurrent environments.
      */
     public void closePool() {
-        while (!connectionPool.isEmpty()) {
-            try {
-                connectionPool.poll().close();
-            } catch (SQLException e) {
-                LOG.error(UNABLE_CLOSE_CONNECTION + ": {}", e.getMessage());
+        synchronized (this) {
```

Comment:
`closePool` method can be executed only when application is closing, there is no sence to make it synchronized.